staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -406,14 +406,38 @@ func (dm *discoveryManager) syncAPIService(apiServiceName string) error {
 	return nil
 }
 
+func (dm *discoveryManager) markInitialStale() {
+	dm.servicesLock.Lock()
+	defer dm.servicesLock.Unlock()
+
+	for key, info := range dm.apiServices {
+		gv := helper.APIServiceNameToGroupVersion(key)
+		entry := apidiscoveryv2beta1.APIVersionDiscovery{
+			Freshness: apidiscoveryv2beta1.DiscoveryFreshnessStale,
+			Version:   gv.Version,
+		}
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, entry)
+		dm.mergedDiscoveryHandler.SetGroupVersionPriority(metav1.GroupVersion(gv), info.groupPriority, info.versionPriority)
+	}
+}
+
 // Spwans a goroutune which waits for added/updated apiservices and updates
 // the discovery document accordingly
-func (dm *discoveryManager) Run(stopCh <-chan struct{}) {
+func (dm *discoveryManager) Run(apiServicesSyncedCh <-chan struct{}, stopCh <-chan struct{}, discoverySyncedCh chan<- struct{}) {
 	klog.Info("Starting ResourceDiscoveryManager")
 
 	// Shutdown the queue since stopCh was signalled
 	defer dm.dirtyAPIServiceQueue.ShutDown()
 
+	// Ensure that apiregistration.k8s.io is the first group in the discovery group.
+	dm.mergedDiscoveryHandler.WithSource(discoveryendpoint.BuiltinSource).SetGroupVersionPriority(APIRegistrationGroupVersion, APIRegistrationGroupPriority, 0)
+
+	// Wait until all apiservices are synced
+	<-apiServicesSyncedCh
```

Comment:
I mentioned that I think this wait should be moved outside of Run, but also note that this should use a `select` with stopCh so that the context is cancellable:

```go
select {
    case <-stopCh:
        // Context cancelled, should abort/clean goroutines
    case <-apiServicesSyncedCh:
        // API services successfully sync
}